### ISSUE TITLE
Feature request: "Argument hints"

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#2201]
Hi,

I've been working on some more completions, and I noticed that sometimes we can't offer a reasonably complete list of options, so I'd like to be able to offer hints instead.

Imagine a program to set your network address, like `ifconfig` or `ip`.

When you enter `ip address add`, you _need_ to enter a valid IP/Netmask after that, and fish/a completion script can know that. What it can't (within reason) is offer all possible addresses with all possible notations. So what it should do instead is just offer the user a hint that an ip address and a netmask are expected.

What I'd like this to look like on the completion script side is an additional option "-H" or "--hint", so that this works:

    complete -c ip -n "__fish_seen_subcommand_from addr; and __fish_seen_subcommand_from add" -H "IP Address/Netmask"

On the UI side, I'm not sure. Like a suggestion, but with a different color? Triggered only after TAB is pressed?

**Comment by ridiculousfish** [comment]
In terms of syntax, we're likely to move to docopt (#478). I'm not sure what the UI could be here either.

**Comment by pickfire** [comment]
> What I'd like this to look like on the completion script side is an additional option "-H" or "--hint", so that this works:

Do you mean this (`-- alias --`)?
![2015-07-15-112941_564x340_scrot](https://cloud.githubusercontent.com/assets/4687791/8690279/e8b5f606-2ae4-11e5-8acc-44c8cf55488f.png)

**Comment by ridiculousfish** [comment]
You can do this via "descriptions," which you can set with a tab delimeter. Example:

    > complete -f -c test_cmd -a '(some_func)'
    > function some_func
            echo "--light"\t"Let there be light"
            echo "-dark"\t"Oh, you think darkness is your ally"
        end

Now tab completing `test_cmd` should show those options with descriptions. Is that what you're looking for?

**Comment by pickfire** [comment]
> Is that what you're looking for?

No, I just don't understand what @faho mean.

**Comment by faho** [comment]
The idea is to show a "hint" to the user that's unrelated to a particular completion.

This is helpful when we can't complete because the space is too big (like ip addresses - no way we can present _all_ possibilities there) - even though this could also be done with regexes, those are quite easy to get wrong and it still doesn't quite tell the user what is expected.

This is also helpful when we know what to complete just to tell the user what it's supposed to mean.

Some examples:

    sleep <TAB>

this expects some kind of time (1s, 5min, 19h....), which of course means the possibilities are endless. 

However, it gets really cool when you look at 

   ln -s <TAB>

Which argument comes first? The path the link points to or the path the link is created at? They're both paths, so we know how to complete them, but it would still be cool if fish did

   ln -s <TAB> (TARGET)

   ln -s /path/to/target <TAB> (LINK)

(maybe without the tab, as I said I'm not sure about the UX)

This might also be included in the manpage parser thingy, since ln's man page says 

     ln [OPTION]... [-T] TARGET LINK_NAME   (1st form)

Of course in the sleep example (or any other option that expects a time, like a "--timeout=" switch) we could complete examples, but that wouldn't work in the ln case and would also often mean the user selects something and then backspaces to change it.

**Comment by pickfire** [comment]
+1, implement that before `zsh` does. Time for *time trial*.

**Comment by ridiculousfish** [comment]
I think I get it! One possibility is to show the hint under the current line. `ln` is a really compelling example.

**Comment by zanchey** [comment]
Yeah, I've been screwing up the arguments to `ln` for about ten years.

**Comment by faho** [comment]
Another case: dd (see https://github.com/fish-shell/fish-shell/issues/752#issuecomment-18070026).

**Comment by floam** [comment]
> I think I get it! One possibility is to show the hint under the current line. ln is a really compelling example.

I think usually you'd want that on the right prompt, unless it really couldn't fit. How about that?

We aren't using those yet. Or at least, not providing any in fish_config.

Hints could display a lot of handy info about stuff under cursor, once you start to think about it. When the command-line is empty they might revert to exit status or whatever it is people tend to do with themes.

**Comment by exploide** [comment]
This would be a really awesome feature. A colleague showed me that zsh apparently already has something like this.

I think this could also lead to a more reasonable use of combining `-a` arguments with a description. Take for example `chsh` and the `-s` options. When completing options, the description "Specify your login shell" is visible and that's completely reasonable:
![chsh-flags](https://user-images.githubusercontent.com/5670236/75886372-65551900-5e28-11ea-9e33-0fa1f8b8905f.png)
But after I decided to indeed use the `-s` flag, the same description is again written for every possible argument:
![chsh-shells](https://user-images.githubusercontent.com/5670236/75886492-96354e00-5e28-11ea-884b-aba41c814087.png)
That's both, space consuming and unnecessary. There are other commands with even more disturbing descriptions.

If we had that hint option, logic could be changed to not show the `-d` description for `-a` arguments. That description would only be shown when completing options. Arguments could either have their tab separated individual description or non at all. Either way that new hint could be shown:

```
~> chsh -s <TAB>    (login shell)
/bin/sh             /bin/bash
/usr/bin/sh         /usr/bin/bash
/usr/bin/tmux       /bin/tmux
/usr/bin/fish       /bin/fish
```

**Comment by EmilySeville7cfg** [comment]
@faho, what about add `-h`/`--hint` option for complete which to show hint for special option when there are too many values can be suggested?

**Comment by faho** [comment]
@EmilySeville7cfg That is what the issue is about.

Someone "simply" has to write it. Unfortunately it's not easy, because *rendering* the hints is tricky, and figuring out when and how exactly they are to be generated is tricky, too.

(also as a sidenote please don't tag people like that, there's no reason to pose this question specifically to us, and the issue is open for everyone)

**Comment by floam** [comment]
Well, the pager is the wrong modality, and we already do something else on the same line for autosuggestions. I think a line below the command being entered, with something that looks like a box pointing at the command line area being hinted about could work.

**Comment by floam** [comment]
I'm kind of imagining this bit of ipython
<img width="598" alt="Screen Shot 2021-11-11 at 11 31 13 AM" src="https://user-images.githubusercontent.com/291142/141357591-0d7ea4d2-97df-4021-af0b-ad904af30bfc.png">
:

**Comment by floam** [comment]
Here's something I did today that could have benefited from hints in a couple ways, but also I think a general UI issue.

I updated the `touch` completions. So `touch -A` takes a specific format and I added a description that does try to help the user get an idea of what kind of input is acceptable. You see this description when you have `touch -` entered:

<img width="334" alt="Screenshot 2022-07-09 at 10 39 58 PM" src="https://user-images.githubusercontent.com/291142/178132877-fb3af74e-03f8-43da-8de9-64343d9e901e.png">

Here's the thing though, not only can we not see the part of that description that really ought to be the hint, the, [-][hh]mm]SS part, after the-A option is typed in and we should be entering that required argument, heck, even the **general description of what -A does** is not visible.

Or let's say you've got `touch -c` typed out and you hit enter - you do not even get an idea of what the heck that option does when you hit enter. You get confusing useless completions of combinations:

**Comment by floam** [comment]
Admittedly, that final example screenshot is probably something that can be improved without extensive `complete`'s abilities at all, just by inserting an initial completion of `-c` into that pager followed by a space with the description for`-c` shown that is able to be selected. It's an oversight.

**Comment by imroc** [comment]
Still not implemented?
